### PR TITLE
[Refactor] bye multiprocessing

### PR DESCRIPTION
**Pull request by shyba**
Removed some optimizations and optimized other slow parts instead.
Main changes:
+ multiprocessing -> nothing (inline decrypt seems fast enough, while multiprocessing was consuming large amounts of RAM for no benefit here)
+ Recursive timed calls -> LoopingCall
+ Sequence from fetched decrypted docs -> sequence from in-memory simple set of indexes
+ Single doc delete from staging -> bulk delete
+ tests for start/stop on decrypter pool.
+ tests for incomplete sequences handling.

### CODE REVIEW

**Comment by meskio** [comment]
Can you add your change to https://github.com/pixelated/soledad/blob/develop/client/changes/next-changelog.rst ?

File: client/src/leap/soledad/client/encdecpool.py [hunk 1]
```diff
@@ -212,29 +174,7 @@ def enqueue_doc_for_encryption(self, doc):
         :param doc: The document to be encrypted.
```
Comment (meskio) [inline client/src/leap/soledad/client/encdecpool.py]: At some point we might need to rename this method, but let's keep it like that for the moment.

File: client/src/leap/soledad/client/encdecpool.py [hunk 2]
```diff
@@ -212,29 +174,7 @@ def enqueue_doc_for_encryption(self, doc):
...
-            while self.running:
-                doc = yield self._encr_queue.get()
-                self._encrypt_doc(doc)
-        except defer.QueueUnderflow:
-            self._delayed_call = reactor.callLater(
-                self.ENCRYPT_LOOP_PERIOD,
-                self._maybe_encrypt_and_recurse)
+        self._encrypt_doc(doc)
 
     def _encrypt_doc(self, doc):
```
Comment (meskio) [inline client/src/leap/soledad/client/encdecpool.py]: As this is not used anywhere besides `enqueue_doc_for_encryption` and this method consist only in a call to `_encrypt_doc` I would move the code of `_encrypt_doc` there.

File: client/src/leap/soledad/client/encdecpool.py [hunk 3]
```diff
@@ -253,9 +193,9 @@ def _encrypt_doc(self, doc):
         secret = self._crypto.secret
         args = doc.doc_id, doc.rev, docstr, key, secret
```
Comment (meskio) [inline client/src/leap/soledad/client/encdecpool.py]: We could use a namedtuple for some of this arguments and reduce the argument list of some of the functions. But this is not really in the scope of this pull-req, I'm just talking loud.

File: client/src/leap/soledad/client/encdecpool.py [hunk 4]
```diff
@@ -507,15 +446,12 @@ def insert_encrypted_received_doc(
         :param idx: The index of this document in the current sync process.
         :type idx: int
```
Comment (meskio) [inline client/src/leap/soledad/client/encdecpool.py]: Maybe we can update the docstring. As far as I understand this method doesn't insert the document for latter decryption.

File: client/src/leap/soledad/client/encdecpool.py (hunk 1 above)
Comment (shyba) [inline client/src/leap/soledad/client/encdecpool.py]: Renamed! I think I got so used to "enqueue" stuff on this module that I didn't realize queue is gone. Also merged with the other method.

File: client/src/leap/soledad/client/encdecpool.py (hunk 4 above)
Comment (shyba) [inline client/src/leap/soledad/client/encdecpool.py]: Updated and merged with async_decrypt_doc. Btw, it still inserts.
Staging on decryption had two inserts, one for encrypted and another for decrypted.
I just jumped directly to the decrypted insert, but it is needed to sort out the sequence as before (but now in memory :))

Comment (shyba) [inline client/src/leap/soledad/client/encdecpool.py]: in memory I mean the calculation, not the insert.

**Comment by shyba** [comment]
Added changes file as well.

File: client/src/leap/soledad/client/encdecpool.py (hunk 3 above)
Comment (shyba) [inline client/src/leap/soledad/client/encdecpool.py]: Yes, but in this specific case I couldn't find a meaningful name for this sequence of parameters.
Without the name it's better to be explicit for now. At least it's just 5 yet.

**Comment by shyba** [comment]
Hey, I removed all comments that claims 7370 solutions and skipped sync_very_large_files test from test_server again.
This refactor made it possible to pass on my machine (8gb ram + SSD), but it won't pass on slower machines. If we enable the test we can break buildbot or snap without a good reason.
7370 is also out of scope for this refactor. Was a good bonus, but not there yet unfortunately.

**Comment by kalikaneko** [comment]
I haven't had time to properly benchmark this. It might be that @andrejb can take this review + benchmark over.

**Comment by kalikaneko** [comment]
Let's try a braindump on this, that has been parked for a while:

1. I wouldn't rush on merging this, since we have 0.8 tag quite close to go with platform.
2. there are several assertions here that need proper measuring. that is a part of the wider _'we need proper benchmarking to aid us in evaluating improvements to performance' _topic. we have some scripts that measure *time* for uploads, but it's cumbersome to use for several reasons:
 (a) it needs to be launched against a known version of the server, that has to be manually upgraded to a certain test branch.
 (b) we haven't resolved on how to integrate the measurements with merge cycles, so that we can have a historic evolution of comparable measures of performance.
 (c) we need to deal with user creation and payload pre-seeding for different batches of tests.

in this case we only want to compare the client with itself, for this a dummy test server might serve well.

> inline decrypt seems fast enough, while multiprocessing was consuming large amounts of RAM for no benefit here

we need to be able to measure this against different payload sizes. I'm afraid that inline decryption of moderately sized payloads will stuck the single core that the reactor is using, with the risk of blocking it from serving other requests.

so, **one metric for benchmarks** should be cpu usage against payload size, but also ability to keep serving certain number of requests per second.

it'd be interesting also to monitor RAM usage. we definitely have some kind of bug there, see https://leap.se/code/issues/7370 . It might be the case that the overhead of serializing the tasks in the current multiprocessing pool implementation is overkill.

3. Implementation details and bugs

it's argued (I'm not sure if explicitely in the changes) that (a) the defer=False option wouldn't work, based on some implementation flaw that makes random order of arrival of download hit ordering bugs.
 
it's also argued that (b) the enc/decr pool is way too complex and that this complexity is too error prone. so, it might be a way or streamlining the programming model behind this module so that we can test it properly w/o involving all the sync machinery.

> Sequence from fetched decrypted docs -> sequence from in-memory simple set of indexes

is this restartable seamlessly after a crash?

> Single doc delete from staging -> bulk delete

I understand also that this brings performance improvements, it'd be good to quantify it.

... 

In general, I assume that the current multiprocessing pool use is not the best of the implementations, but it's there to alleviate load that we saw in the client when doing massive syncs, blocking the reactor ability to serve other requests. I think the way forward would be to find a more frictionless way of dealing with delegating tasks to a pool (ampoule was a good candidate). 

that said, if measurements show us that the cpu usage for realistic payloads is not too bad, and that the tradeoff is good enough by using way less RAM, i'm ok with merging the changes until we can work on a more suitable multi-core solution.

**Comment by kalikaneko** [comment]
in case it's useful, some experiments I've been doing to understand better the 3 scenarios in the discussion:
https://github.com/kalikaneko/soledad-perf

I hope to be able to extend the prototypes to handle soledad-client and gnugp payloads anytime soon.

don't hesitate to ping me with any ongoing progress on this, this is an exciting topic!

**Comment by kalikaneko** [comment]
#320 has been merged

**Pull request closed without merging**